File: tests/integration/check_p4_toolchain.py
"""Offline staging/preflight safety checks; no third-party Python packages."""
import importlib.util
import json
from pathlib import Path
import tempfile
import unittest
from unittest.mock import patch
from types import SimpleNamespace
import subprocess

ROOT = Path(__file__).resolve().parents[2]
spec = importlib.util.spec_from_file_location("toolchain", ROOT / "tools/toolchain.py")
module = importlib.util.module_from_spec(spec)
spec.loader.exec_module(module)


class ToolchainTests(unittest.TestCase):
    def test_installed_records_require_exact_set_build_and_hash(self):
        with tempfile.TemporaryDirectory() as tmp:
            prefix = Path(tmp)
            (prefix / "conda-meta").mkdir()
            record = {"name": "pkg", "version": "1", "build": "a", "sha256": "a" * 64}
            file = prefix / "conda-meta/pkg.json"
            file.write_text(json.dumps(record))
            module.check_records(prefix, [record])
            for key, value in (("version", "2"), ("build", "b"), ("sha256", "b" * 64)):
                file.write_text(json.dumps(dict(record, **{key: value})))
                with self.assertRaises(ValueError):
                    module.check_records(prefix, [record])
            file.write_text(json.dumps(record))
            (prefix / "conda-meta/extra.json").write_text(json.dumps(dict(record, name="extra")))
            with self.assertRaises(ValueError):
                module.check_records(prefix, [record])

    def test_lock_consistency_checks_url_and_checksum(self):
        record = {"url": "https://example.org/pkg.conda", "md5": "a" * 32}
        with tempfile.TemporaryDirectory() as tmp:
            lock = Path(tmp) / "lock.txt"
            lock.write_text("@EXPLICIT\n" + record["url"] + "#" + record["md5"] + "\n")
            module.check_lock(lock, [record])
            lock.write_text("@EXPLICIT\n" + record["url"] + "#" + "b" * 32 + "\n")
            with self.assertRaises(ValueError):
                module.check_lock(lock, [record])

    def test_existing_destination_and_dangling_link_rejected(self):
        with tempfile.TemporaryDirectory() as tmp:
            root = Path(tmp)
            with self.assertRaises(ValueError):
                module.new_destination(root)
            link = root / "link"
            link.symlink_to(root / "missing")
            with self.assertRaises(ValueError):
                module.new_destination(link)
            self.assertEqual(module.new_destination(root / "new"), root / "new")

    def test_failed_staging_propagates_without_verifying_prefix(self):
        with tempfile.TemporaryDirectory() as tmp:
            args = SimpleNamespace(prefix=Path(tmp) / "space ; literal", kind="alignment", mamba="mamba")
            with patch.object(module.shutil, "which", return_value="/tools/mamba"), \
                 patch.object(module.subprocess, "run", side_effect=subprocess.CalledProcessError(7, "mamba")) as run, \
                 patch.object(module, "check_records") as verified:
                with self.assertRaises(subprocess.CalledProcessError):
                    module.stage(args)
                argv = run.call_args.args[0]
                self.assertEqual(argv[argv.index("--prefix") + 1], str(args.prefix))
                self.assertNotIn("shell", run.call_args.kwargs)
                verified.assert_not_called()

    def test_r_environment_and_auxiliary_threads_are_constrained(self):
        with patch.dict(module.os.environ, {"R_LIBS_USER": "/untrusted", "LD_LIBRARY_PATH": "/wrong", "OMP_NUM_THREADS": "999"}):
            env = module.environment()
        self.assertEqual(env["R_LIBS_USER"], module.os.devnull)
        self.assertNotIn("LD_LIBRARY_PATH", env)
        self.assertEqual(env["OMP_NUM_THREADS"], "1")


if __name__ == "__main__":
    unittest.main()

File: tests/integration/check_p4_toolchain.R
#!/usr/bin/env Rscript
source(file.path(dirname(sub('^--file=','',grep('^--file=',commandArgs(),value=TRUE)[1])),'../../tools/toolchain.R'))
root<-new_directory('toolchain-tests-')
fails<-function(expr)stopifnot(inherits(tryCatch({force(expr);NULL},error=identity),'error'))
dir.create(file.path(root,'conda-meta')); record<-list(name='pkg',version='1',build='a',sha256=strrep('a',64)); path<-file.path(root,'conda-meta/pkg.json')
write_json(record,path); check_records(root,list(record))
for(key in c('version','build','sha256')) {changed<-record; changed[[key]]<-'changed';write_json(changed,path);fails(check_records(root,list(record)))}
write_json(record,path); extra<-record;extra$name<-'extra';write_json(extra,file.path(root,'conda-meta/extra.json'));fails(check_records(root,list(record)))
lock<-file.path(root,'lock'); expected<-list(list(url='https://example.org/pkg.conda',md5=strrep('a',32)))
writeLines(c('@EXPLICIT',paste0(expected[[1]]$url,'#',expected[[1]]$md5)),lock);check_lock(lock,expected)
writeLines(c('@EXPLICIT',paste0(expected[[1]]$url,'#',strrep('b',32))),lock);fails(check_lock(lock,expected))
fails(new_destination(root));file.symlink(file.path(root,'absent'),file.path(root,'link'));fails(new_destination(file.path(root,'link')));stopifnot(new_destination(file.path(root,'new'))==file.path(root,'new'))
stub<-file.path(root,'mamba');writeLines(c('#!/bin/sh','printf "%s\\n" "$@" > "$P7_ARGS"','exit 7'),stub);Sys.chmod(stub,'0755')
with_environment(list(P7_ARGS=file.path(root,'args')),fails(stage_toolchain('alignment',file.path(root,'space ; literal'),stub)))
argv<-readLines(file.path(root,'args'));stopifnot(argv[match('--prefix',argv)+1L]==file.path(root,'space ; literal'),!file.exists(file.path(root,'space ; literal')))
with_environment(list(R_LIBS_USER='/untrusted',LD_LIBRARY_PATH='/wrong',OMP_NUM_THREADS='999'),with_environment(clean_environment(),stopifnot(Sys.getenv('R_LIBS_USER')=='/dev/null',Sys.getenv('LD_LIBRARY_PATH')=='',Sys.getenv('OMP_NUM_THREADS')=='1')))
# MD5-explicit mamba records can omit SHA256 and retain a legacy cache suffix.
# A verified cached archive supplies SHA256; absent or modified archives fail.
unlink(file.path(root,'conda-meta/extra.json'))
archive<-file.path(root,'pkg.conda');writeLines('pinned archive content',archive)
pinned<-c(record,list(url='https://example.org/pkg.conda',md5=unname(tools::md5sum(archive))))
pinned$sha256<-sha256(archive);installed<-pinned;installed$sha256<-NULL
installed$package_tarball_full_path<-file.path(root,'pkg.tar.bz2');write_json(installed,path)
check_records(root,list(pinned))
writeLines('modified archive',archive);fails(check_records(root,list(pinned)))
unlink(archive);fails(check_records(root,list(pinned)))
unlink(root,recursive=TRUE);cat('PASS exact package/lock records, absent output, literal failed staging, clean environment\n')
